src/lib.rs

```diff
@@ -57,226 +57,306 @@
 //! }
 //!
 //! impl ObjectProvider for MyProvider {
-//!     fn provide<'a>(&'a self, request: Pin<&mut Request<'a>>) {
+//!     fn provide<'a>(&'a self, request: &mut Request<'a>) {
 //!         request
 //!             .provide_ref::<PathBuf>(&self.path)
 //!             .provide_ref::<Path>(&self.path)
-//!             .provide_ref::<dyn Debug>(&self.path);
+//!             .provide_ref::<dyn Debug>(&self.path)
+//!             .provide_value_with::<String, _>(|| {
+//!                 self.path.to_string_lossy().into_owned()
+//!             });
 //!     }
 //! }
 //! ```
 
 use core::any::TypeId;
-use core::fmt;
-use core::marker::{PhantomData, PhantomPinned};
-use core::pin::Pin;
+use core::marker::PhantomData;
 
-struct ReqRef<T: ?Sized + 'static>(&'static T);
-struct ReqVal<T: 'static>(T);
+mod private {
+    pub trait Response {}
+    impl<T> Response for Option<T> {}
+}
+
+/// When used by `Request`, `TypeId::of::<RefMarker<T>>()` indicates that
+/// `response` is of type `Option<&'a T>`.
+struct RefMarker<T: ?Sized + 'static>(T);
+
+/// When used by `Request`, `TypeId::of::<ValueMarker<T>>()` indicates that
+/// `response` is of type `Option<T>`.
+struct ValueMarker<T: 'static>(T);
 
 /// A dynamic request for an object based on its type.
-#[repr(C)]
-pub struct Request<'a> {
+pub struct Request<'a, R = dyn private::Response + 'a>
+where
+    R: ?Sized + private::Response,
+{
+    marker: PhantomData<&'a ()>,
```

Comment:
I was worried about this not being invariant, but... `&mut Request<'a>` *is* invariant over `'a` because of the `&mut`, as it's basically a `&mut (TypeId, Option<&'a T>)` for some unknown `T`, so it has to be invariant to avoid allowing writing in a shorter lifetime.